File: 121/cpp/main.cpp
#include <algorithm>
#include <iostream>
#include <vector>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        int max_profit = 0;
        int min_price = prices[0];

        for (int& current_price : prices) {
            max_profit = std::max(max_profit, current_price - min_price);
            min_price = std::min(min_price, current_price);
        }

        return max_profit;
    }
};

int main() {
    Solution solution;

    std::vector<int> prices1 = {7,1,5,3,6,4};
    std::cout << solution.maxProfit(prices1) << "\n";

    std::vector<int> prices2 = {7,6,4,3,1};
    std::cout << solution.maxProfit(prices2) << "\n";

    return 0;
}

File: 121/python3/main.py
#!/usr/bin/env python3

class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        max_profit, min_price = 0, prices[0]

        for current_price in prices:
            max_profit = max(max_profit, current_price - min_price)
            min_price = min(min_price, current_price)

        return max_profit

solution = Solution()

print(solution.maxProfit([7,1,5,3,6,4]))
print(solution.maxProfit([7,6,4,3,1]))
